python/mlir_structured/dialects/indexing.py

```diff
@@ -0,0 +1,112 @@
+# Copyright 2021 The IREE Authors
+#
+# Licensed under the Apache License v2.0 with LLVM Exceptions.
+# See https://llvm.org/LICENSE.txt for license information.
+# SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+import re
+from typing import Union, Tuple
+
+from . import arith as arith_dialect
+from . import linalg
+from . import tensor as tensor_dialect
+# noinspection PyUnresolvedReferences
+from ._indexing_ops_gen import *
+from .linalg.opdsl.lang.emitter import _BodyBuilder, _is_floating_point_type, _is_integer_type, \
+  _is_index_type, _is_complex_type
+# noinspection PyUnresolvedReferences
+from .._mlir_libs import _mlirStructuredPasses as _cextStructuredPasses
+from .._mlir_libs._structuredDialects.indexing import *
+from ..ir import Value, Type, RankedTensorType, ShapedType, Operation, IndexType
+
+res_val_reg = re.compile(r"(%\w+) =")
+
+_body_builder = _BodyBuilder({}, {}, {})
+
+
+class Scalar(ArithValue):
```

Comment:
This and the next class are (amazingly elegant!) syntactic sugar for creating MLIR. Please document the high-level idea of doing so in a paragraph or two, probably at the beginning of this file.